<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An R package for aggregating single-cell RNA-Seq data and metadata in an ExpressionSet object, along with biomaRt gene annotation and basic cell filtering â€¢ scprep</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="An R package for aggregating single-cell RNA-Seq data and metadata in an ExpressionSet object, along with biomaRt gene annotation and basic cell filtering">
<meta property="og:description" content="This R package aggregates single-cell RNA-Seq data and metadata in an ExpressionSet object, along with biomaRt gene annotation and basic cell filtering.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">scprep</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/vignette_scprep_usage.html">scprep: Single-cell Data Object Types Guide</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div class="float">
<img src="inst/extdata/scprep_Logo_v2.png" alt="Logo"><div class="figcaption">Logo</div>
</div>
<hr>
<div class="section level1">
<div class="page-header"><h1 id="scprep">
<strong>scprep</strong><a class="anchor" aria-label="anchor" href="#scprep"></a>
</h1></div>
<p>An R package for aggregating single-cell RNA-Seq data and metadata in ExpressionSet, Seurat, or SingleCellExperiment objects, along with biomaRt gene annotation, basic cell filtering and QC metric calculation.</p>
<p>Data that may be used for testing <strong>scprep</strong> may be accessed via the <strong><a href="https://github.com/g-duclos/scdata" class="external-link">scdata</a></strong> package.</p>
<p>For post-processing &amp; secondary analysis of data that has been ingested by <strong>scprep</strong>, this package may be coupled with <strong><a href="https://github.com/g-duclos/scpost" class="external-link">scpost</a></strong>.</p>
<hr>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<div class="section level3">
<h3 id="option-1-install-from-github">Option 1: Install from GitHub<a class="anchor" aria-label="anchor" href="#option-1-install-from-github"></a>
</h3>
<p>The <strong>scprep</strong> R package can be installed from Github using devtools:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"g-duclos/scprep"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="option-2-docker-container-recommended">Option 2: Docker Container (Recommended)<a class="anchor" aria-label="anchor" href="#option-2-docker-container-recommended"></a>
</h3>
<p>For a containerized environment with all dependencies pre-installed:</p>
<p><strong>Quick Start with Docker:</strong></p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co"># Build the Docker image</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="ex">docker</span> build <span class="at">-t</span> scprep .</span>
<span id="cb2-3"><a href="#cb2-3" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" tabindex="-1"></a><span class="co"># Run interactive R session with scprep</span></span>
<span id="cb2-5"><a href="#cb2-5" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-it</span> <span class="at">--rm</span> scprep</span>
<span id="cb2-6"><a href="#cb2-6" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" tabindex="-1"></a><span class="co"># Run with data directory mounted</span></span>
<span id="cb2-8"><a href="#cb2-8" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">-it</span> <span class="at">--rm</span> <span class="at">-v</span> /path/to/your/data:/scprep/data scprep</span></code></pre></div>
<p><strong>Using Docker Compose:</strong></p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Start the scprep container</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="ex">docker-compose</span> up scprep</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="co"># For RStudio interface (optional)</span></span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a><span class="ex">docker-compose</span> up scprep-rstudio</span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co"># Then navigate to http://localhost:8788</span></span></code></pre></div>
<p><strong>Docker Image Features:</strong> - Pre-installed R 4.2.1+ with all required dependencies - Seurat v5, Matrix, Biobase, biomaRt, and other dependencies configured - Optimized for single-cell RNA-Seq workflows - Volume mounting for data input/output</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="documentation">Documentation<a class="anchor" aria-label="anchor" href="#documentation"></a>
</h2>
<p>ğŸ“– <strong><a href="https://htmlpreview.github.io/?https://github.com/g-duclos/scprep/blob/main/vignettes/vignette_scprep_usage.html" class="external-link">Complete Usage Guide &amp; Object Types Vignette</a></strong> - Comprehensive guide demonstrating how to read 10X data, convert between object types, and explore data structure</p>
<hr>
</div>
<div class="section level2">
<h2 id="getting-started">Getting Started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<div class="section level4">
<h4 id="annotation">Annotation<a class="anchor" aria-label="anchor" href="#annotation"></a>
</h4>
<p>Specify the sample metadata - view the annotation file here: <a href="inst/extdata/scprep_annotation.csv">scprep_annotation.csv</a> (click â€œRaw fileâ€ at the top right to download) * Column 1: â€œSample_IDâ€ corresponds to the name of each sample * Column 2: â€œIndexâ€ corresponds to the name of the PCR index used during library preparation * Column 3: â€œSample_Projectâ€ corresponds to the name of the project affiliated with all samples * Column 4: â€œReferenceâ€ corresponds to the name of reference used when running Cell Ranger * Additional columns can be added corresponding to useful metadata for a particular experiment</p>
</div>
<div class="section level4">
<h4 id="parameters">Parameters<a class="anchor" aria-label="anchor" href="#parameters"></a>
</h4>
<p>Specify the pipeline parameters - view the parameters file here: <a href="inst/extdata/scprep_parameters.csv">scprep_parameters.csv</a> (click â€œRaw fileâ€ at the top right to download)</p>
<p><strong>Critical Parameters</strong> * <em>dir_input</em> (path to input directory) * <em>dir_output</em> (path to output directory) * <em>file_type</em> (input format: â€œh5â€ or â€œmtxâ€) * <em>output_type</em> (object type: â€œesetâ€, â€œseuratâ€, or â€œsceâ€) * <em>gene_id</em> (gene identifier type: â€œensemblâ€ or â€œsymbolâ€)</p>
</div>
<div class="section level4">
<h4 id="input-directory">Input Directory<a class="anchor" aria-label="anchor" href="#input-directory"></a>
</h4>
<p>Define the input directory (<em>dir_input</em>), which must contain a subdirectory named after each sample. The input directory (<em>dir_input</em>) must also contain the following file(s) produced by 10X Genomicsâ€™ <em>Cell Ranger</em> pipeline:</p>
<p><strong>Option 1: H5 Format (Recommended)</strong> * <strong>filtered_feature_bc_matrix.h5</strong> (the filtered gene counts matrix for each sample)</p>
<p><strong>Option 2: MTX Format</strong> * <strong>matrix.mtx</strong> (gene expression count matrix in Matrix Market format) * <strong>barcodes.tsv</strong> (cell barcodes) * <strong>genes.tsv</strong> or <strong>features.tsv</strong> (gene identifiers and names)</p>
<p><strong>Additional Files:</strong> * If working with 10x Genomics Immune Profiling assay that includes 5â€™ RNA-Seq with TCR or Ig V(D)J data: <strong>filtered_contig_annotations.csv</strong></p>
<p><strong>Format Selection:</strong> Set the <code>file_type</code> parameter in <code>scprep_parameters.csv</code>: - <code>file_type="h5"</code> for H5 format (supports multimodal data like CITE-seq/ATAC-seq) - <code>file_type="mtx"</code> for MTX format (RNA expression data only)</p>
</div>
<div class="section level4">
<h4 id="dependencies">Dependencies:<a class="anchor" aria-label="anchor" href="#dependencies"></a>
</h4>
<p><strong>scprep</strong> requires the following R packages: - <strong>Seurat</strong> (&gt;= 5.0.0): For reading 10X Genomics data formats and Seurat object creation - <strong>Matrix</strong> (&gt;= 1.2-0): For sparse matrix operations<br>
- <strong>Biobase</strong>: For ExpressionSet data structures - <strong>SingleCellExperiment</strong>: For SingleCellExperiment object creation - <strong>R</strong> (&gt;= 4.0.0): Minimum R version</p>
<p><strong>Note:</strong> All dependencies are automatically installed when using <code><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">devtools::install_github()</a></code> or the Docker container.</p>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="output-object-types">Output Object Types<a class="anchor" aria-label="anchor" href="#output-object-types"></a>
</h2>
<p><strong>scprep</strong> now supports three different output object types:</p>
<div class="section level3">
<h3 id="expressionset-default">
<strong>ExpressionSet</strong> (Default)<a class="anchor" aria-label="anchor" href="#expressionset-default"></a>
</h3>
<ul>
<li>Traditional Bioconductor S4 object for storing expression data</li>
<li>Metadata stored in <code><a href="https://rdrr.io/pkg/Biobase/man/phenoData.html" class="external-link">pData()</a></code> and <code><a href="https://rdrr.io/pkg/Biobase/man/featureData.html" class="external-link">fData()</a></code> slots</li>
<li>Compatible with Bioconductor workflows</li>
<li>Maintains full backward compatibility</li>
</ul>
</div>
<div class="section level3">
<h3 id="seurat-v5">
<strong>Seurat</strong> (v5)<a class="anchor" aria-label="anchor" href="#seurat-v5"></a>
</h3>
<ul>
<li>Popular single-cell analysis framework object</li>
<li>Metadata stored in <code>meta.data</code> slot</li>
<li>Ready for Seurat downstream analysis workflows</li>
<li>Supports multi-modal data (protein, VDJ)</li>
</ul>
</div>
<div class="section level3">
<h3 id="singlecellexperiment">
<strong>SingleCellExperiment</strong><a class="anchor" aria-label="anchor" href="#singlecellexperiment"></a>
</h3>
<ul>
<li>Modern Bioconductor S4 object for single-cell data</li>
<li>Metadata stored in <code>colData()</code> slot</li>
<li>Compatible with Bioconductor/scater workflows</li>
<li>Supports alternative experiments (protein data)</li>
</ul>
<hr>
</div>
</div>
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>Template function to aggregate gene counts matrices from multiple samples, store aggregated counts matrices and metadata (for samples and genes) in an ExpressionSet, Seurat, or SingleCellExperiment object, add biomaRt gene annotation, perform cell filtering, and calculate select QC metrics.</p>
<p><strong>Output type is specified in the <code>scprep_parameters.csv</code> file:</strong></p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biobase" class="external-link">Biobase</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Output type determined by output_type parameter in scprep_parameters.csv</span></span>
<span><span class="va">dataset</span> <span class="op">&lt;-</span> <span class="fu">scprep</span><span class="fu">::</span><span class="fu"><a href="reference/template_scprep.html">template_scprep</a></span><span class="op">(</span>dir_output<span class="op">=</span><span class="va">dir_output</span><span class="op">)</span></span></code></pre></div>
<p><strong>Core functions:</strong></p>
<details><summary>
Read filtered_feature_matrix_bc.h5 file for each sample listed in scprep_annotation.csv into an ExpressionSet, Seurat, or SingleCellExperiment object. Add sample metadata from scprep_annotation.csv to the object metadata. Calculate transcripts (â€œUMIsâ€) per cell and genes (â€œGenesâ€) per cell (&gt;=1 transcript detected) and add to the object metadata. If working with multi-modal RNA/V(D)J, CITE, or RNA/ATAC data, this function will also store the V(D)J, CITE ADT surface protein, or ATAC information in the object.
</summary><pre>
# Build Seurat object with RNA data (no additional modalities)
dataset 
</pre></details><details><summary>
Add gene-level metadata to the object (Ensembl ID, gene ID, chromosome #, chromosome start, chromosome stop, biotype) using the <a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html" class="external-link">biomaRt R package</a>. Calculate transcripts per cell derived from each biotype and chromosome and add to the objectâ€™s cell-level metadata slot.
</summary><pre>
# Add gene-level metadata to the object
dataset 
</pre></details><details><summary>
Assign barcodes the status of â€œDeadâ€ if a high percentage of total transcripts are derived from mitochondrial genes (â€œmax_mitoâ€ fraction specified in â€œscprep_parameters.csvâ€). Assign barcodes the status of â€œDebrisâ€ if low numbers of total transcripts or genes were detected per cell (â€œmin_umiâ€ and â€œmin_geneâ€ specified in â€œscprep_parameters.csvâ€). Assign barcodes the status of â€œCellâ€ if percent mitochondrial transcripts is less than â€œmax_mitoâ€ and if total transcripts is greater than â€œmin_umiâ€.
</summary><pre>
# Assign status of high quality "Cell", "Dead", or "Debris" to each barcode and add to object metadata
dataset 
</pre></details><details><summary>
Additional features
</summary><ul>
<li>
Store parameters specified in the scprep_parameters.csv file in a list labeled â€œParametersâ€ in the â€œassayDataâ€ slot labeled â€œParamsâ€ in the ExpressionSet object
</li>
<li>
Generate random seeds utilized for this analysis and store in a list labeled â€œSeedsâ€ in the â€œassayDataâ€ slot labeled â€œParamsâ€ in the ExpressionSet object
</li>
<li>
Select genes with at least 3 transcript counts in a pre-specified (see â€œgene_filterâ€ in scprep_parameters.csv) percentage (default = 0.1%) of cells and label as â€œExpressedâ€ in â€œfDataâ€ slot of ExpressionSet object. All other genes are labeled as â€œNot_Expressedâ€ in â€œfDataâ€ slot of ExpressionSet object
</li>
<li>
Save ExpressionSet RDS file in <em>dir_output</em>
</li>
</ul></details><hr>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3" class="external-link">GPL-3</a></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing scprep</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Grant Duclos <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/YOUR-ORCID-ID" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Grant Duclos.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
